### PR TITLE
Copy the class source validator from WG.

### PR DESCRIPTION
**Pull request by wizjany**
Warns when other plugins incorrectly shade WE or our libs.

### CODE REVIEW

File: worldedit-bukkit/src/main/java/com/sk89q/bukkit/util/ClassSourceValidator.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Validates that certain specified classes came from the same source as
+ * a plugin.
+ */
+public class ClassSourceValidator {
+
+    private static final Logger log = Logger.getLogger(ClassSourceValidator.class.getCanonicalName());
```
Comment (octylFractal) [inline worldedit-bukkit/src/main/java/com/sk89q/bukkit/util/ClassSourceValidator.java:42]:
```suggestion
    private static final Logger LOGGER = Logger.getLogger(ClassSourceValidator.class);
```

File: worldedit-bukkit/src/main/java/com/sk89q/bukkit/util/ClassSourceValidator.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    private static final String separatorLine = Strings.repeat("*", 46);
```
Comment (octylFractal) [inline worldedit-bukkit/src/main/java/com/sk89q/bukkit/util/ClassSourceValidator.java:43]:
```suggestion
    private static final String SEPARATOR_LINE = Strings.repeat("*", 46);
```

File: worldedit-bukkit/src/main/java/com/sk89q/bukkit/util/ClassSourceValidator.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+     *
+     * @param classes A list of classes to check
+     * @return The results
+     */
+    public Map<Class<?>, CodeSource> findMismatches(List<Class<?>> classes) {
+        checkNotNull(classes, "classes");
+
+        Map<Class<?>, CodeSource> mismatches = Maps.newHashMap();
+
+        if (expectedCodeSource != null) {
```
Comment (octylFractal) [inline worldedit-bukkit/src/main/java/com/sk89q/bukkit/util/ClassSourceValidator.java:71]: Check this before `mismatches` is initialized and return `ImmutableMap.of`?

File: worldedit-bukkit/src/main/java/com/sk89q/bukkit/util/ClassSourceValidator.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+     *
+     * @param classes The list of classes to check
+     */
+    public void reportMismatches(List<Class<?>> classes) {
+        if (Boolean.getBoolean("enginehub.disable.class.validation")) {
+            return;
+        }
+        Map<Class<?>, CodeSource> mismatches = findMismatches(classes);
+
+        if (!mismatches.isEmpty()) {
```
Comment (octylFractal) [inline worldedit-bukkit/src/main/java/com/sk89q/bukkit/util/ClassSourceValidator.java:96]: Why not `if (empty) return`? seems like a waste to indent this block of code

File: worldedit-bukkit/src/main/java/com/sk89q/bukkit/util/ClassSourceValidator.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+            StringBuilder builder = new StringBuilder("\n");
+
+            builder.append(separatorLine).append("\n");
+            builder.append("** /!\\    SEVERE WARNING    /!\\\n");
+            builder.append("** \n");
+            builder.append("** A plugin developer has copied and pasted a portion of \n");
```
Comment (octylFractal) [inline worldedit-bukkit/src/main/java/com/sk89q/bukkit/util/ClassSourceValidator.java:102]:
```suggestion
            builder.append("** A plugin developer has included a portion of \n");
```
I don't think we should assume copy/paste, shading is likely too (and while still technically copy/paste, it's not quite the same).

File: worldedit-bukkit/src/main/java/com/sk89q/worldedit/bukkit/WorldEditPlugin.java [hunk 6]
```diff
@@ -139,6 +143,12 @@ public void onLoad() {
      */
     @Override
     public void onEnable() {
+
+        // Catch bad things being done by naughty plugins that include
+        // WorldEdit's classes
+        ClassSourceValidator verifier = new ClassSourceValidator(this);
+        verifier.reportMismatches(ImmutableList.of(World.class, CommandManager.class));
```
Comment (octylFractal) [inline worldedit-bukkit/src/main/java/com/sk89q/worldedit/bukkit/WorldEditPlugin.java:150]: Maybe toss in `EditSession` here too?

Comment (me4502) [reply worldedit-bukkit/src/main/java/com/sk89q/worldedit/bukkit/WorldEditPlugin.java:150]: Maybe Actor too

**Pull request merged**